cmd/mailbox-controller/main.go

```diff
@@ -308,65 +332,65 @@ func (ctl *mbCtl) sync1(ctx context.Context, ref any) {
 		return false
 	}
-	cluster := logicalcluster.Name(wsNameParts[0])
-	targetName := wsNameParts[1]
+	cluster := ref.cluster
+	targetName := ref.name
 	syncTarget, err := ctl.synctargetLister.Cluster(cluster).Get(targetName)
 	if err != nil && !k8sapierrors.IsNotFound(err) {
-		logger.Error(err, "Unable to Get referenced SyncTarget", "wsNameParts", wsNameParts)
+		logger.Error(err, "Unable to Get referenced SyncTarget", "ref", ref)
 		return true
 	}
+	wsName := ref.mailboxWSName()
 	workspace, err := ctl.workspaceLister.Get(wsName)
 	if err != nil && !k8sapierrors.IsNotFound(err) {
-		logger.Error(err, "Unable to Get referenced Workspace", "wsName", wsName)
+		logger.Error(err, "Unable to Get referenced Workspace", "ref", ref, "wsName", wsName)
 		return true
 	}
 	if syncTarget == nil || syncTarget.DeletionTimestamp != nil {
 		if workspace == nil || workspace.DeletionTimestamp != nil {
-			logger.V(3).Info("Both absent or deleting, nothing to do", "wsName", wsName)
+			logger.V(3).Info("Both absent or deleting, nothing to do", "ref", ref)
 			return false
 		}
 		err := ctl.workspaceClient.Delete(ctx, wsName, metav1.DeleteOptions{Preconditions: &metav1.Preconditions{UID: &workspace.UID}})
 		if err == nil || k8sapierrors.IsNotFound(err) {
-			logger.V(2).Info("Deleted unwanted workspace", "wsName", wsName)
+			logger.V(2).Info("Deleted unwanted workspace", "ref", ref)
 			return false
 		}
-		logger.Error(err, "Failed to delete unwanted workspace", "wsName", wsName)
+		logger.Error(err, "Failed to delete unwanted workspace", "ref", ref)
 		return true
 	}
 	// Now we have established that the SyncTarget exists and is not being deleted
 	if workspace == nil {
 		ws := &tenancyv1alpha1.Workspace{
 			ObjectMeta: metav1.ObjectMeta{
-				Name: wsName,
+				Annotations: map[string]string{SyncTargetNameAnnotationKey: ref.name},
+				Name:        wsName,
 			},
 			Spec: tenancyv1alpha1.WorkspaceSpec{},
 		}
 		_, err := ctl.workspaceClient.Create(ctx, ws, metav1.CreateOptions{FieldManager: "mailbox-controller"})
-		if err != nil {
-			logger.Error(err, "Failed to create workspace", "wsName", wsName)
-			return true
+		if err == nil {
+			logger.V(2).Info("Created missing workspace", "ref", ref)
+			return false
 		}
-		logger.V(2).Info("Created missing workspace", "wsName", wsName)
-		return false
-
+		if k8sapierrors.IsAlreadyExists(err) {
+			logger.V(3).Info("Missing workspace was created concurrently", "ref", ref)
+			return false
+		}
+		logger.Error(err, "Failed to create workspace", "ref", ref)
+		return true
 	}
 	if workspace.DeletionTimestamp != nil {
-		logger.V(3).Info("Wanted workspace is being deleted, will retry later", "wsName", wsName)
+		logger.V(3).Info("Wanted workspace is being deleted, will retry later", "ref", ref)
 		return true
 	}
-	logger.V(3).Info("Both exist and are not being deleted, nothing to do", "wsName", wsName)
+	logger.V(3).Info("Both exist and are not being deleted, nothing to do", "ref", ref)
```

Comment:
Same as above comment: Can we explicitly say "Both syncTarget & workspace are ...."

Reply:
Done.